variantlib/plugins/py_envs.py

```diff
@@ -0,0 +1,341 @@
+
+    elif "posix_local" in scheme_names:
+        # The Python that ships on Debian/Ubuntu varies the default scheme to
+        # install to /usr/local
+        # But it does not (yet) set the "venv" scheme.
+        # If we're the Debian "posix_local" scheme is available, but "venv"
+        # is not, we use "posix_prefix" instead which is venv-compatible there.
+        paths = sysconfig.get_paths(scheme="posix_prefix", vars=config_vars)
+
+    elif "osx_framework_library" in scheme_names:
+        # The Python that ships with the macOS developer tools varies the
+        # default scheme depending on whether the ``sys.prefix`` is part of a framework.
+        # But it does not (yet) set the "venv" scheme.
+        # If the Apple-custom "osx_framework_library" scheme is available but "venv"
+        # is not, we use "posix_prefix" instead which is venv-compatible there.
+        paths = sysconfig.get_paths(scheme="posix_prefix", vars=config_vars)
+
+    else:
+        paths = sysconfig.get_paths(vars=config_vars)
+
+    executable = os.path.join(  # noqa: PTH118
+        paths["scripts"], "python.exe" if os.name == "nt" else "python"
+    )
+    if not os.path.exists(executable):  # noqa: PTH110
+        msg = f"Virtual environment creation failed, executable {executable} missing"
+        raise RuntimeError(msg)
+
+    return executable, paths["scripts"], paths["purelib"]
+
+
+@functools.cache
+def _fs_supports_symlink() -> bool:
+    """Return True if symlinks are supported"""
+    # Using definition used by venv.main()
+    if os.name != "nt":
+        return True
+
+    # Windows may support symlinks (setting in Windows 10)
+    with tempfile.NamedTemporaryFile(prefix="variant-symlink-") as tmp_file:
+        dest = f"{tmp_file}-b"
+        try:
+            os.symlink(tmp_file.name, dest)
+            os.unlink(dest)  # noqa: PTH108
+        except (OSError, NotImplementedError, AttributeError):
+            return False
+        return True
+
+
+class BasePythonEnv(abc.ABC):
+    @abc.abstractmethod
+    def __enter__(self) -> Self:
+        """
+        Enter the environment.
+        """
+
+    @abc.abstractmethod
+    def __exit__(self, *args: object) -> None:
+        """
+        Exit the environment.
+        """
+
+
+class IsolatedPythonEnvMixin:
+    _venv_path: pathlib.Path | None = None
+
+    @property
+    def python_executable(self) -> pathlib.Path:
+        return self._get_venv_path(0)
+
+    @property
+    def script_dir(self) -> pathlib.Path:
+        return self._get_venv_path(1)
+
+    @property
+    def package_dir(self) -> pathlib.Path:
+        return self._get_venv_path(2)
+
+    def _get_venv_path(self, idx: int) -> pathlib.Path:
+        if self._venv_path is None or not self._venv_path.exists():
+            raise FileNotFoundError
+        assert 0 <= idx <= 2
+        return pathlib.Path(_find_executable_and_scripts(str(self._venv_path))[idx])
```

Comment:
I don't get this at all. Why aren't we just calling `_find_executable_and_scripts()` in `__init__()` and setting them all as class attributes?

Reply:
Oh because the venv is not created when the environment is created but rather when you enter the contextmanager.

Maybe this could be done differently - feel free to experiment